src/java/main/org/apache/zookeeper/client/StaticHostProvider.java

```diff
@@ -87,15 +86,69 @@ public StaticHostProvider(Collection<InetSocketAddress> serverAddresses)
         Collections.shuffle(this.serverAddresses);
     }
 
+    /**
+     * In Java 7, we have a method getHostString, but earlier versions do not support it.
+     * This method is to provide a replacement for InetSocketAddress.getHostString().
+     *
+     * It evaluates to a hostname if one is available and otherwise it returns the
+     * string representation of the IP address.
+     *
+     * @param addr
+     * @return Hostname string of address parameter
+     */
+    private String getHostString(InetSocketAddress addr) {
+        String hostString;
+        InetAddress ia = addr.getAddress();
+
+        if (ia != null) {
+            // If the string starts with '/', then it has no hostname
+            // and we want to avoid the reverse lookup, so we return
+            // the string representation of the address.
+            if (ia.toString().startsWith("/")) {
+                hostString = ia.getHostAddress();
+            } else {
+                hostString = addr.getHostName();
+            }
+        } else {
+            // According to the Java 6 documentation, if the hostname is
+            // unresolved, then the string before the colon is the hostname.
+            String addrString = addr.toString();
+            hostString = addrString.substring(0, addrString.lastIndexOf(':'));
+        }
+
+        return hostString;
+    }
+
     public int size() {
         return serverAddresses.size();
     }
 
     public InetSocketAddress next(long spinDelay) {
-        ++currentIndex;
-        if (currentIndex == serverAddresses.size()) {
-            currentIndex = 0;
+        // Handle possible connection error by re-resolving hostname if possible
+        if (!connectedSinceNext) {
+            InetSocketAddress curAddr = serverAddresses.get(currentIndex);
+            if (!curAddr.getHostString().equals(curAddr.getAddress().getHostAddress())) {
+                try {
+                    int thePort = curAddr.getPort();
+                    InetAddress resolvedAddresses[] = InetAddress.getAllByName(getHostString(curAddr));
+                    if (resolvedAddresses.length == 1) {
+                        serverAddresses.set(currentIndex, new InetSocketAddress(resolvedAddresses[0], thePort));
+                    } else {
+                        serverAddresses.remove(currentIndex);
+                        for (InetAddress resolvedAddress : resolvedAddresses) {
+                            InetSocketAddress newAddr = new InetSocketAddress(resolvedAddress, thePort);
+                            if (!serverAddresses.contains(newAddr)) {
+                                serverAddresses.add(newAddr);
+                            }
+                        }
```

Comment:
We shuffle initially to avoid having all clients connecting to the same server in the case they are all given the same connect string. If the array of addresses has already been shuffled (in the constructor), then the order followed in this method will be the shuffled one. I don't see a strong reason for re-shuffling, as we are not bringing it back to the original order by resolving again.